src/server/pfs/server/driver_compact.go

```diff
@@ -12,188 +11,74 @@ import (
-		if err != nil {
-			return err
-		}
-		renewer.Add(res.OutputPath)
-		return d.storage.Copy(ctx, res.OutputPath, outputPath, 0)
-	})
-}
-
-type compactSpec struct {
-	master     *work.Master
-	inputPaths []string
-	maxFanIn   int
-}
-
-type compactResult struct {
-	OutputPath string
-}
-
-// compactIter is one level of compaction.  It will only perform compaction
-// if len(inputPaths) <= params.maxFanIn otherwise it will split inputPaths recursively.
-func (d *driver) compactIter(ctx context.Context, params compactSpec) (*compactResult, error) {
-	if len(params.inputPaths) <= params.maxFanIn {
-		return d.shardedCompact(ctx, params.master, params.inputPaths)
-	}
-	childSize := params.maxFanIn
-	for len(params.inputPaths)/childSize > params.maxFanIn {
-		childSize *= params.maxFanIn
-	}
-	// TODO: use an errgroup to make the recursion concurrecnt.
-	// this requires changing the master to allow multiple calls to RunSubtasks
-	// don't forget to pass the errgroups childCtx to compactIter instead of ctx.
-	var res *compactResult
-	if err := d.storage.WithRenewer(ctx, defaultTTL, func(ctx context.Context, renewer *renew.StringSet) error {
-		var childOutputPaths []string
-		for start := 0; start < len(params.inputPaths); start += childSize {
-			end := start + childSize
-			if end > len(params.inputPaths) {
-				end = len(params.inputPaths)
-			}
-			res, err := d.compactIter(ctx, compactSpec{
-				master:     params.master,
-				inputPaths: params.inputPaths[start:end],
-				maxFanIn:   params.maxFanIn,
-			})
-			if err != nil {
-				return err
-			}
-			renewer.Add(res.OutputPath)
-			childOutputPaths = append(childOutputPaths, res.OutputPath)
-		}
-		var err error
-		res, err = d.shardedCompact(ctx, params.master, childOutputPaths)
-		return err
-	}); err != nil {
-		return nil, err
-	}
-	return res, nil
-}
-
-// shardedCompact generates shards for the fileset(s) in inputPaths,
-// gives those shards to workers, and waits for them to complete.
-// Fan in is bound by len(inputPaths), concatenating shards have
-// fan in of one because they are concatenated sequentially.
-func (d *driver) shardedCompact(ctx context.Context, master *work.Master, inputPaths []string) (*compactResult, error) {
-	scratch := path.Join(tmpRepo, uuid.NewWithoutDashes())
-	compaction := &pfs.Compaction{InputPrefixes: inputPaths}
-	var subtasks []*work.Task
-	var shardOutputs []string
-	fs, err := d.storage.Open(ctx, inputPaths)
-	if err != nil {
-		return nil, err
-	}
-	if err := d.storage.Shard(ctx, fs, func(pathRange *index.PathRange) error {
-		shardOutputPath := path.Join(scratch, strconv.Itoa(len(subtasks)))
-		shard, err := serializeShard(&pfs.Shard{
-			Compaction: compaction,
+func (d *driver) compact(master *work.Master, ids []fileset.ID) (*fileset.ID, error) {
+	// serialize access to RunSubtasks, the compactor may call workerFunc concurrently
+	mu := sync.Mutex{}
+	workerFunc := func(ctx context.Context, task fileset.CompactionTask) (*fileset.ID, error) {
```

Comment:
We probably should change this function to take a set of tasks. Right now we are not processing shards in parallel and we kind of cannot because the work package expects you to not have parallel `RunSubtasks` calls. The expected usage is to provide a batch, or to setup a channel that you eventually close for a batch.

Reply:
I'm guessing you set things up like this to make keeping the order of shards simpler?

Reply:
I mostly wanted to express the compaction concurrently in the fileset package.  And then impose the one call constraint here, hence the Mutex.  Is there really that much of a performance benefit to batching these tasks?  I was hoping we could change the work package to allow single calls concurrently.

My prior model for a work coordination package is that tasks are given an identifier and there is essentially a pub-sub channel setup for that identifier, the callers wait for the callee, or timeout.
It's trivial to layer caching on top of that, and that is a great way to get checkpointing, without complicated the algorithm.  There is no state for a job, only work skipping based on cached computations.

Reply:
Added `CompactionBatchWorker` and switched back to the old model. Left a TODO next to `CompactionWorker` in the fileset package.